File: macos/Services/AI/Auth/ChatGPTOAuthService.swift
// ChatGPTOAuthService.swift
// Gridex
//
// Owns the full ChatGPT OAuth lifecycle:
//   - signIn(providerId:)         — interactive PKCE flow, opens the browser
//   - tokenBundle(providerId:)    — non-interactive, refreshes if access token
//                                    is within `refreshSkew` of expiring
//   - signOut(providerId:)        — clears the Keychain blob
//   - currentStatus(providerId:)  — UI helper, no network
//
// Concurrency model: actor. In-flight refreshes are coalesced per providerId so
// that a burst of `tokenBundle()` calls only triggers one network round-trip.

import AppKit
import Foundation
import os

actor ChatGPTOAuthService {

    private static let log = Logger(subsystem: "com.gridex.gridex", category: "ChatGPTOAuth")

    private let keychain: KeychainServiceProtocol
    private let urlSession: URLSession
    private var refreshInFlight: [UUID: Task<ChatGPTTokenBundle, Error>] = [:]

    init(keychainService: KeychainServiceProtocol, urlSession: URLSession = .shared) {
        self.keychain = keychainService
        self.urlSession = urlSession
    }

    enum SignInStatus: Equatable, Sendable {
        case signedOut
        case signedIn(email: String?, plan: String?)
    }

    // MARK: - Public API

    /// Runs the full PKCE flow: opens browser, waits for /auth/callback,
    /// exchanges the code for tokens, persists them, returns the bundle.
    /// Call from any context — interaction with NSWorkspace happens on MainActor
    /// internally.
    @discardableResult
    func signIn(providerId: UUID) async throws -> ChatGPTTokenBundle {
        Self.log.info("ChatGPT OAuth sign-in starting for provider \(providerId.uuidString, privacy: .public)")

        let verifier  = PKCE.makeVerifier()
        let challenge = PKCE.challenge(for: verifier)
        let state     = PKCE.makeState()

        let server = ChatGPTOAuthLoopbackServer()
        let port = try await server.start()
        let redirectURI = "http://localhost:\(port)\(ChatGPTOAuthConstants.callbackPath)"

        // Open the browser. NSWorkspace must be touched on the main thread.
        let authURL = buildAuthorizeURL(challenge: challenge, state: state, redirectURI: redirectURI)
        Self.log.debug("Opening browser to authorize URL")
        let didOpenBrowser = await MainActor.run {
            NSWorkspace.shared.open(authURL)
        }
        guard didOpenBrowser else {
            server.stop()
            Self.log.error("ChatGPT OAuth sign-in failed: LaunchServices could not open browser")
            throw GridexError.aiProviderError("Could not open browser for ChatGPT sign-in")
        }

        // Block on the browser callback (or 3-min timeout).
        let callback: ChatGPTOAuthCallback
        do {
            callback = try await server.awaitCallback(timeout: ChatGPTOAuthConstants.callbackTimeout)
        } catch {
            server.stop()
            Self.log.error("OAuth callback failed: \(error.localizedDescription, privacy: .public)")
            throw GridexError.aiProviderError("Sign-in failed: \(error.localizedDescription)")
        }

        // CSRF check — the auth server echoes our `state` back to the redirect URI.
        guard callback.state == state else {
            Self.log.error("OAuth state mismatch — possible CSRF")
            throw GridexError.aiProviderError("State mismatch — sign-in aborted (possible CSRF)")
        }

        // Exchange the authorization code for tokens.
        let tokens = try await exchangeCodeForTokens(
            code: callback.code,
            verifier: verifier,
            redirectURI: redirectURI
        )

        let bundle = try buildBundle(from: tokens)
        try keychain.saveChatGPTTokens(providerId: providerId, bundle: bundle)
        Self.log.info("ChatGPT OAuth sign-in complete; email=\(bundle.email ?? "?", privacy: .private)")

        // Bring Gridex back to the front — Safari may not auto-close the tab.
        await MainActor.run {
            NSApp.activate(ignoringOtherApps: true)
        }

        return bundle
    }

    /// Returns a usable token bundle, refreshing via `refresh_token` if the
    /// stored `access_token` is within `refreshSkew` of expiring.
    /// Coalesces concurrent calls per `providerId`.
    func tokenBundle(providerId: UUID) async throws -> ChatGPTTokenBundle {
        guard let bundle = try keychain.loadChatGPTTokens(providerId: providerId) else {
            throw GridexError.aiAPIKeyMissing
        }

        let exp = JWTDecoder.expiration(of: bundle.accessToken) ?? .distantPast
        if Date() < exp.addingTimeInterval(-ChatGPTOAuthConstants.refreshSkew) {
            return bundle
        }

        if let inflight = refreshInFlight[providerId] {
            return try await inflight.value
        }

        let task = Task<ChatGPTTokenBundle, Error> { [self] in
            try await self.performRefresh(bundle: bundle, providerId: providerId)
        }
        refreshInFlight[providerId] = task
        defer { refreshInFlight[providerId] = nil }
        return try await task.value
    }

    /// Deletes the persisted bundle. Cheap; does not call the auth server.
    func signOut(providerId: UUID) throws {
        Self.log.info("ChatGPT sign-out for \(providerId.uuidString, privacy: .public)")
        try keychain.deleteChatGPTTokens(providerId: providerId)
    }

    /// Non-network status used by the settings sheet to render "Signed in as ..."
    /// chrome.
    func currentStatus(providerId: UUID) -> SignInStatus {
        guard let bundle = try? keychain.loadChatGPTTokens(providerId: providerId) else {
            return .signedOut
        }
        return .signedIn(email: bundle.email, plan: bundle.planType)
    }

    // MARK: - Internal

    private struct TokenResponse: Decodable {
        let access_token: String
        let refresh_token: String?
        let id_token: String?
    }

    private func buildAuthorizeURL(challenge: String, state: String, redirectURI: String) -> URL {
        var comps = URLComponents(url: ChatGPTOAuthConstants.issuer.appendingPathComponent("oauth/authorize"),
                                  resolvingAgainstBaseURL: false)!
        comps.queryItems = [
            URLQueryItem(name: "response_type", value: "code"),
            URLQueryItem(name: "client_id", value: ChatGPTOAuthConstants.clientId),
            URLQueryItem(name: "redirect_uri", value: redirectURI),
            URLQueryItem(name: "scope", value: ChatGPTOAuthConstants.scopes),
            URLQueryItem(name: "code_challenge", value: challenge),
            URLQueryItem(name: "code_challenge_method", value: "S256"),
            URLQueryItem(name: "state", value: state),
            URLQueryItem(name: "id_token_add_organizations", value: "true"),
            URLQueryItem(name: "codex_cli_simplified_flow", value: "true"),
            URLQueryItem(name: "originator", value: ChatGPTOAuthConstants.originator),
        ]
        return comps.url!
    }

    private func exchangeCodeForTokens(code: String, verifier: String, redirectURI: String) async throws -> TokenResponse {
        let url = ChatGPTOAuthConstants.issuer.appendingPathComponent("oauth/token")
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")

        // form-encode body
        let body = [
            "grant_type":    "authorization_code",
            "code":          code,
            "redirect_uri":  redirectURI,
            "client_id":     ChatGPTOAuthConstants.clientId,
            "code_verifier": verifier,
        ]
            .map { "\($0.key)=\(encode($0.value))" }
            .joined(separator: "&")
        req.httpBody = Data(body.utf8)

        let (data, response) = try await urlSession.data(for: req)
        guard let http = response as? HTTPURLResponse else {
            throw GridexError.aiProviderError("Token exchange: non-HTTP response")
        }
        guard (200..<300).contains(http.statusCode) else {
            let detail = parseErrorBody(data)
            Self.log.error("Token exchange failed: HTTP \(http.statusCode) \(detail, privacy: .public)")
            throw GridexError.aiProviderError("Token exchange failed (HTTP \(http.statusCode)): \(detail)")
        }

        do {
            return try JSONDecoder().decode(TokenResponse.self, from: data)
        } catch {
            Self.log.error("Token exchange: malformed body — \(error.localizedDescription, privacy: .public)")
            throw GridexError.aiProviderError("Token exchange returned an unexpected payload")
        }
    }

    /// JSON-bodied refresh per Codex CLI behaviour. The auth server returns the
    /// new access_token (and may rotate refresh_token + id_token). When fields
    /// are absent we keep the old values.
    private func performRefresh(bundle: ChatGPTTokenBundle, providerId: UUID) async throws -> ChatGPTTokenBundle {
        Self.log.info("ChatGPT token refresh for \(providerId.uuidString, privacy: .public)")
        let url = ChatGPTOAuthConstants.issuer.appendingPathComponent("oauth/token")
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: String] = [
            "client_id":     ChatGPTOAuthConstants.clientId,
            "grant_type":    "refresh_token",
            "refresh_token": bundle.refreshToken,
        ]
        req.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await urlSession.data(for: req)
        guard let http = response as? HTTPURLResponse else {
            throw GridexError.aiProviderError("Refresh: non-HTTP response")
        }

        if (200..<300).contains(http.statusCode) {
            let parsed = try JSONDecoder().decode(TokenResponse.self, from: data)
            // Merge new values over the old bundle. refresh_token / id_token may rotate.
            var updated = bundle
            updated.accessToken = parsed.access_token
            if let rt = parsed.refresh_token, !rt.isEmpty { updated.refreshToken = rt }
            if let idt = parsed.id_token, !idt.isEmpty {
                updated.idToken = idt
                applyClaims(into: &updated)
            }
            updated.obtainedAt = Date()
            try keychain.saveChatGPTTokens(providerId: providerId, bundle: updated)
            Self.log.info("Refresh ok; new exp=\(JWTDecoder.expiration(of: updated.accessToken)?.timeIntervalSince1970 ?? 0)")
            return updated
        }

        // Failure path. If the auth server says the refresh_token is no longer
        // valid, blow the bundle away and surface aiAPIKeyMissing — the UI will
        // prompt the user to sign in again.
        let detail = parseErrorBody(data)
        let lower = detail.lowercased()
        let dead = ["refresh_token_expired", "refresh_token_reused", "refresh_token_invalidated"]
        if dead.contains(where: lower.contains) {
            Self.log.notice("Refresh token rejected (\(detail, privacy: .public)); clearing keychain")
            try? keychain.deleteChatGPTTokens(providerId: providerId)
            throw GridexError.aiAPIKeyMissing
        }
        throw GridexError.aiProviderError("Refresh failed (HTTP \(http.statusCode)): \(detail)")
    }

    /// Decodes the id_token claims into the bundle's UI-friendly fields.
    private func applyClaims(into bundle: inout ChatGPTTokenBundle) {
        guard let claims = try? JWTDecoder.payload(of: bundle.idToken) else { return }
        bundle.email     = claims["email"] as? String ?? bundle.email
        bundle.accountId = claims["chatgpt_account_id"] as? String ?? bundle.accountId
        bundle.planType  = claims["chatgpt_plan_type"] as? String ?? bundle.planType
    }

    /// Build a fresh bundle from a successful token exchange response.
    private func buildBundle(from tokens: TokenResponse) throws -> ChatGPTTokenBundle {
        guard let idToken = tokens.id_token, !idToken.isEmpty else {
            throw GridexError.aiProviderError("Token exchange returned no id_token")
        }
        guard let refresh = tokens.refresh_token, !refresh.isEmpty else {
            throw GridexError.aiProviderError("Token exchange returned no refresh_token (offline_access scope missing?)")
        }
        var b = ChatGPTTokenBundle(
            accessToken: tokens.access_token,
            refreshToken: refresh,
            idToken: idToken
        )
        applyClaims(into: &b)
        return b
    }

    private func encode(_ value: String) -> String {
        // Application/x-www-form-urlencoded — URL-encode + use `+` for spaces.
        var allowed = CharacterSet.urlQueryAllowed
        allowed.remove(charactersIn: "+&=")
        return value.addingPercentEncoding(withAllowedCharacters: allowed) ?? value
    }

    private func parseErrorBody(_ data: Data) -> String {
        if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            if let err = json["error"] as? String {
                if let desc = json["error_description"] as? String { return "\(err): \(desc)" }
                return err
            }
            if let msg = json["message"] as? String { return msg }
        }
        return String(data: data, encoding: .utf8)?
            .trimmingCharacters(in: .whitespacesAndNewlines) ?? "(empty body)"
    }
}

File: linux/src/Services/AI/Auth/ChatGPTOAuthService.cpp
#include "Services/AI/Auth/ChatGPTOAuthService.h"

#include <QByteArray>
#include <QNetworkAccessManager>
#include <QProcess>
#include <QStringList>
#include <QNetworkReply>
#include <QNetworkRequest>
#include <QUrl>
#include <QUrlQuery>

#include <nlohmann/json.hpp>

#include "Data/Keychain/SecretStore.h"
#include "Services/AI/Auth/ChatGPTOAuthConstants.h"
#include "Services/AI/Auth/JwtDecoder.h"
#include "Services/AI/Auth/OAuthLoopbackServer.h"
#include "Services/AI/Auth/PKCE.h"

namespace gridex {

namespace {

QString sv(std::string_view s) {
    return QString::fromUtf8(s.data(), static_cast<int>(s.size()));
}

QString parseErrorBody(const QByteArray& data) {
    try {
        auto j = nlohmann::json::parse(data.toStdString());
        if (j.contains("error")) {
            QString err = QString::fromStdString(j["error"].get<std::string>());
            if (j.contains("error_description")) {
                return err + ": " + QString::fromStdString(j["error_description"].get<std::string>());
            }
            return err;
        }
        if (j.contains("message")) {
            return QString::fromStdString(j["message"].get<std::string>());
        }
    } catch (...) {}
    return QString::fromUtf8(data).trimmed();
}

}  // namespace

ChatGPTOAuthService::ChatGPTOAuthService(SecretStore* secretStore, QObject* parent)
    : QObject(parent),
      secretStore_(secretStore),
      nam_(std::make_unique<QNetworkAccessManager>()) {}

ChatGPTOAuthService::~ChatGPTOAuthService() = default;

void ChatGPTOAuthService::signIn(const QString& providerKey) {
    using namespace oauth;

    // Tear down any previous attempt's listener.
    server_.reset(new OAuthLoopbackServer(this));
    std::uint16_t port = server_->start();
    if (port == 0) {
        emit signInFailed(providerKey, tr("Could not bind any loopback port for OAuth callback"));
        return;
    }

    QString verifier  = PKCE::makeVerifier();
    QString challenge = PKCE::challengeFor(verifier);
    QString state     = PKCE::makeState();
    QString redirectUri =
        QString("http://localhost:%1%2").arg(port).arg(sv(kCallbackPath));

    QUrl authorize(sv(kIssuer) + sv(kAuthorizePath));
    QUrlQuery q;
    q.addQueryItem("response_type",                 "code");
    q.addQueryItem("client_id",                     sv(kClientId));
    q.addQueryItem("redirect_uri",                  redirectUri);
    q.addQueryItem("scope",                         sv(kScopes));
    q.addQueryItem("code_challenge",                challenge);
    q.addQueryItem("code_challenge_method",         "S256");
    q.addQueryItem("state",                         state);
    q.addQueryItem("id_token_add_organizations",    "true");
    q.addQueryItem("codex_cli_simplified_flow",     "true");
    q.addQueryItem("originator",                    sv(kOriginator));
    authorize.setQuery(q);

    // Open the user's default browser via xdg-open. Avoids the Qt6::Gui
    // dependency on this service library (services link Qt6::Network only).
    if (!QProcess::startDetached("xdg-open", {authorize.toString(QUrl::FullyEncoded)})) {
        emit signInFailed(providerKey, tr("Could not open browser (xdg-open failed)"));
        server_.reset();
        return;
    }

    server_->awaitCallback(
        kCallbackTimeoutSeconds,
        [this, providerKey, state, verifier, redirectUri](OAuthCallback cb) {
            onCallbackSuccess(providerKey, state, verifier, redirectUri, cb.code, cb.state);
        },
        [this, providerKey](QString msg) {
            onCallbackError(providerKey, msg);
        });
}

void ChatGPTOAuthService::onCallbackSuccess(const QString& providerKey,
                                            const QString& expectedState,
                                            const QString& verifier,
                                            const QString& redirectUri,
                                            const QString& code,
                                            const QString& callbackState) {
    if (callbackState != expectedState) {
        emit signInFailed(providerKey,
            tr("OAuth state mismatch — sign-in aborted (possible CSRF)"));
        return;
    }
    exchangeCodeForTokens(providerKey, code, verifier, redirectUri);
}

void ChatGPTOAuthService::onCallbackError(const QString& providerKey, const QString& message) {
    emit signInFailed(providerKey, message);
}

void ChatGPTOAuthService::exchangeCodeForTokens(const QString& providerKey,
                                                 const QString& code,
                                                 const QString& verifier,
                                                 const QString& redirectUri) {
    using namespace oauth;
    QUrl tokenUrl(sv(kIssuer) + sv(kTokenPath));
    QNetworkRequest req(tokenUrl);
    req.setHeader(QNetworkRequest::ContentTypeHeader,
                  "application/x-www-form-urlencoded");
    req.setAttribute(QNetworkRequest::RedirectPolicyAttribute,
                     QNetworkRequest::NoLessSafeRedirectPolicy);

    QUrlQuery body;
    body.addQueryItem("grant_type",    "authorization_code");
    body.addQueryItem("code",          code);
    body.addQueryItem("redirect_uri",  redirectUri);
    body.addQueryItem("client_id",     sv(kClientId));
    body.addQueryItem("code_verifier", verifier);
    QByteArray payload = body.toString(QUrl::FullyEncoded).toUtf8();

    QNetworkReply* reply = nam_->post(req, payload);
    connect(reply, &QNetworkReply::finished, this, [this, reply, providerKey]() {
        reply->deleteLater();
        int status = reply->attribute(QNetworkRequest::HttpStatusCodeAttribute).toInt();
        QByteArray data = reply->readAll();
        if (status < 200 || status >= 300) {
            emit signInFailed(providerKey,
                tr("Token exchange failed (HTTP %1): %2")
                    .arg(status).arg(parseErrorBody(data)));
            return;
        }
        try {
            auto j = nlohmann::json::parse(data.toStdString());
            ChatGPTTokenBundle bundle;
            bundle.accessToken  = j.value("access_token", "");
            bundle.refreshToken = j.value("refresh_token", "");
            bundle.idToken      = j.value("id_token", "");
            if (bundle.accessToken.empty() || bundle.refreshToken.empty()) {
                emit signInFailed(providerKey,
                    tr("Token exchange returned no access_token / refresh_token"));
                return;
            }
            applyClaims(bundle);
            persistBundle(providerKey, bundle);
            emit signInCompleted(providerKey, bundle);
        } catch (const std::exception& e) {
            emit signInFailed(providerKey,
                tr("Token exchange returned an unexpected payload: ") + e.what());
        }
    });
}

void ChatGPTOAuthService::requestFreshToken(const QString& providerKey) {
    auto bundle = currentBundle(providerKey);
    if (!bundle) {
        emit signInFailed(providerKey, tr("Not signed in"));
        return;
    }
    auto exp = oauth::JwtDecoder::expiration(QString::fromStdString(bundle->accessToken));
    bool stale = !exp || (std::chrono::system_clock::now() +
                          std::chrono::seconds(oauth::kRefreshSkewSeconds) >= *exp);
    if (!stale) {
        emit tokenRefreshed(providerKey, *bundle);
        return;
    }
    {
        std::lock_guard lk(mu_);
        if (refreshInFlight_) return;  // coalesce; the in-flight call will emit
        refreshInFlight_ = true;
    }
    performRefresh(providerKey, *bundle);
}

void ChatGPTOAuthService::performRefresh(const QString& providerKey,
                                         ChatGPTTokenBundle bundle) {
    using namespace oauth;
    QUrl tokenUrl(sv(kIssuer) + sv(kTokenPath));
    QNetworkRequest req(tokenUrl);
    req.setHeader(QNetworkRequest::ContentTypeHeader, "application/json");
    req.setAttribute(QNetworkRequest::RedirectPolicyAttribute,
                     QNetworkRequest::NoLessSafeRedirectPolicy);
    nlohmann::json body = {
        {"client_id",     std::string(kClientId)},
        {"grant_type",    "refresh_token"},
        {"refresh_token", bundle.refreshToken},
    };
    QByteArray payload = QByteArray::fromStdString(body.dump());

    QNetworkReply* reply = nam_->post(req, payload);
    connect(reply, &QNetworkReply::finished, this,
            [this, reply, providerKey, bundle]() mutable {
        reply->deleteLater();
        {
            std::lock_guard lk(mu_);
            refreshInFlight_ = false;
        }
        int status = reply->attribute(QNetworkRequest::HttpStatusCodeAttribute).toInt();
        QByteArray data = reply->readAll();

        if (status >= 200 && status < 300) {
            try {
                auto j = nlohmann::json::parse(data.toStdString());
                bundle.accessToken = j.value("access_token", bundle.accessToken);
                if (j.contains("refresh_token") && j["refresh_token"].is_string()) {
                    auto rt = j["refresh_token"].get<std::string>();
                    if (!rt.empty()) bundle.refreshToken = rt;
                }
                if (j.contains("id_token") && j["id_token"].is_string()) {
                    auto idt = j["id_token"].get<std::string>();
                    if (!idt.empty()) {
                        bundle.idToken = idt;
                        applyClaims(bundle);
                    }
                }
                bundle.obtainedAt = std::chrono::system_clock::now();
                persistBundle(providerKey, bundle);
                emit tokenRefreshed(providerKey, bundle);
            } catch (...) {
                emit signInFailed(providerKey,
                    tr("Refresh returned an unexpected payload"));
            }
            return;
        }

        // Refresh-token-rejected family: clear the bundle, prompt re-sign-in.
        QString detail = parseErrorBody(data).toLower();
        const QStringList dead = {"refresh_token_expired",
                                  "refresh_token_reused",
                                  "refresh_token_invalidated"};
        for (const auto& d : dead) {
            if (detail.contains(d)) {
                if (secretStore_) secretStore_->deleteChatGPTTokens(providerKey.toStdString());
                emit signInFailed(providerKey,
                    tr("Sign-in expired — please sign in with ChatGPT again"));
                return;
            }
        }
        emit signInFailed(providerKey,
            tr("Refresh failed (HTTP %1): %2").arg(status).arg(parseErrorBody(data)));
    });
}

void ChatGPTOAuthService::persistBundle(const QString& providerKey,
                                        const ChatGPTTokenBundle& bundle) {
    if (!secretStore_) return;
    std::string raw = bundle.toJson().dump();
    secretStore_->saveChatGPTTokens(providerKey.toStdString(), raw);
}

void ChatGPTOAuthService::applyClaims(ChatGPTTokenBundle& bundle) {
    if (bundle.idToken.empty()) return;
    try {
        auto claims = oauth::JwtDecoder::payload(QString::fromStdString(bundle.idToken));
        if (claims.contains("email") && claims["email"].is_string()) {
            bundle.email = claims["email"].get<std::string>();
        }
        if (claims.contains("chatgpt_account_id") && claims["chatgpt_account_id"].is_string()) {
            bundle.accountId = claims["chatgpt_account_id"].get<std::string>();
        }
        if (claims.contains("chatgpt_plan_type") && claims["chatgpt_plan_type"].is_string()) {
            bundle.planType = claims["chatgpt_plan_type"].get<std::string>();
        }
    } catch (...) {}
}

std::optional<ChatGPTTokenBundle>
ChatGPTOAuthService::currentBundle(const QString& providerKey) const {
    if (!secretStore_) return std::nullopt;
    auto raw = secretStore_->loadChatGPTTokens(providerKey.toStdString());
    if (!raw) return std::nullopt;
    try {
        auto j = nlohmann::json::parse(*raw);
        return ChatGPTTokenBundle::fromJson(j);
    } catch (...) {
        return std::nullopt;
    }
}

void ChatGPTOAuthService::signOut(const QString& providerKey) {
    if (secretStore_) secretStore_->deleteChatGPTTokens(providerKey.toStdString());
}

ChatGPTSignInStatus ChatGPTOAuthService::status(const QString& providerKey) const {
    ChatGPTSignInStatus s;
    auto b = currentBundle(providerKey);
    if (!b) return s;
    s.signedIn = true;
    if (b->email)    s.email    = QString::fromStdString(*b->email);
    if (b->planType) s.planType = QString::fromStdString(*b->planType);
    return s;
}

}  // namespace gridex
